File: backend/app/schemas/health.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, field_validator

from app.models.deworming import DewormingType
from app.models.routine import RoutineType


# ---------------- Weight ----------------

class WeightCreate(BaseModel):
    weight_kg: Decimal
    recorded_at: date

    @field_validator("weight_kg")
    @classmethod
    def validate_weight(cls, v: Decimal) -> Decimal:
        if v <= 0 or v > Decimal("200"):
            raise ValueError("体重必须在 0-200kg 之间")
        # At most 2 decimal places
        if v.as_tuple().exponent < -2:
            raise ValueError("体重最多保留两位小数")
        return v

    @field_validator("recorded_at")
    @classmethod
    def validate_date(cls, v: date) -> date:
        if v > date.today():
            raise ValueError("记录日期不能是未来日期")
        return v


class WeightUpdate(WeightCreate):
    pass


class WeightResponse(BaseModel):
    id: int
    pet_id: int
    user_id: int
    weight_kg: Decimal
    recorded_at: date
    created_at: datetime

    model_config = {"from_attributes": True}


class WeightListResponse(BaseModel):
    weights: list[WeightResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


# ---------------- Deworming ----------------

class DewormingCreate(BaseModel):
    deworming_type: DewormingType
    dewormed_at: date

    @field_validator("dewormed_at")
    @classmethod
    def validate_date(cls, v: date) -> date:
        if v > date.today():
            raise ValueError("驱虫日期不能是未来日期")
        return v


class DewormingUpdate(DewormingCreate):
    pass


class DewormingResponse(BaseModel):
    id: int
    pet_id: int
    user_id: int
    deworming_type: DewormingType
    dewormed_at: date
    created_at: datetime

    model_config = {"from_attributes": True}


class DewormingListResponse(BaseModel):
    dewormings: list[DewormingResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class DewormingCycleUpdate(BaseModel):
    internal_cycle_days: int | None = None
    external_cycle_days: int | None = None
    combined_cycle_days: int | None = None
    internal_reminder_enabled: bool | None = None
    external_reminder_enabled: bool | None = None
    combined_reminder_enabled: bool | None = None

    @field_validator(
        "internal_cycle_days",
        "external_cycle_days",
        "combined_cycle_days",
    )
    @classmethod
    def validate_cycle(cls, v: int | None) -> int | None:
        if v is not None and (v < 1 or v > 365):
            raise ValueError("驱虫周期必须在 1-365 天之间")
        return v


class DewormingCycleResponse(BaseModel):
    internal_cycle_days: int | None
    external_cycle_days: int | None
    combined_cycle_days: int | None
    internal_reminder_enabled: bool
    external_reminder_enabled: bool
    combined_reminder_enabled: bool


class DewormingStatusItem(BaseModel):
    reminder_enabled: bool
    last_dewormed_at: date | None
    cycle_days: int | None
    next_due_at: date | None
    days_remaining: int | None
    is_overdue: bool | None


class DewormingStatusResponse(BaseModel):
    internal: DewormingStatusItem
    external: DewormingStatusItem
    combined: DewormingStatusItem


# ---------------- Vaccination ----------------

class VaccinationCreate(BaseModel):
    vaccine_type: str
    vaccinated_at: date

    @field_validator("vaccine_type")
    @classmethod
    def validate_type(cls, v: str) -> str:
        v = v.strip()
        if not v or len(v) > 100:
            raise ValueError("疫苗类型长度为1-100个字符")
        return v

    @field_validator("vaccinated_at")
    @classmethod
    def validate_date(cls, v: date) -> date:
        if v > date.today():
            raise ValueError("接种日期不能是未来日期")
        return v


class VaccinationUpdate(VaccinationCreate):
    pass


class VaccinationResponse(BaseModel):
    id: int
    pet_id: int
    user_id: int
    vaccine_type: str
    vaccinated_at: date
    created_at: datetime

    model_config = {"from_attributes": True}


class VaccinationListResponse(BaseModel):
    vaccinations: list[VaccinationResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class VaccineTypePresetResponse(BaseModel):
    preset_types: list[str]
    pet_type: str


# ---------------- Routine ----------------

class RoutineCreate(BaseModel):
    routine_type: RoutineType
    performed_at: date

    @field_validator("performed_at")
    @classmethod
    def validate_date(cls, v: date) -> date:
        if v > date.today():
            raise ValueError("日常记录日期不能是未来日期")
        return v


class RoutineUpdate(RoutineCreate):
    pass


class RoutineResponse(BaseModel):
    id: int
    pet_id: int
    user_id: int
    routine_type: RoutineType
    performed_at: date
    created_at: datetime

    model_config = {"from_attributes": True}


class RoutineListResponse(BaseModel):
    routines: list[RoutineResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class RoutineCycleUpdate(BaseModel):
    bath_cycle_days: int | None = None
    nail_trim_cycle_days: int | None = None
    grooming_cycle_days: int | None = None
    bath_reminder_enabled: bool | None = None
    nail_trim_reminder_enabled: bool | None = None
    grooming_reminder_enabled: bool | None = None

    @field_validator(
        "bath_cycle_days",
        "nail_trim_cycle_days",
        "grooming_cycle_days",
    )
    @classmethod
    def validate_cycle(cls, v: int | None) -> int | None:
        if v is not None and (v < 1 or v > 365):
            raise ValueError("日常周期必须在 1-365 天之间")
        return v


class RoutineCycleResponse(BaseModel):
    bath_cycle_days: int | None
    nail_trim_cycle_days: int | None
    grooming_cycle_days: int | None
    bath_reminder_enabled: bool
    nail_trim_reminder_enabled: bool
    grooming_reminder_enabled: bool


class RoutineStatusItem(BaseModel):
    reminder_enabled: bool
    last_performed_at: date | None
    cycle_days: int | None
    next_due_at: date | None
    days_remaining: int | None
    is_overdue: bool | None


class RoutineStatusResponse(BaseModel):
    bath: RoutineStatusItem
    nail_trim: RoutineStatusItem
    grooming: RoutineStatusItem

File: frontend/lib/models/health.dart
class WeightRecord {
  final int id;
  final int petId;
  final int userId;
  final double weightKg;
  final String recordedAt;
  final String createdAt;

  const WeightRecord({
    required this.id,
    required this.petId,
    required this.userId,
    required this.weightKg,
    required this.recordedAt,
    required this.createdAt,
  });

  factory WeightRecord.fromJson(Map<String, dynamic> json) {
    final raw = json['weight_kg'];
    final value = raw is num ? raw.toDouble() : double.parse(raw.toString());
    return WeightRecord(
      id: json['id'] as int,
      petId: json['pet_id'] as int,
      userId: json['user_id'] as int,
      weightKg: value,
      recordedAt: json['recorded_at'] as String,
      createdAt: json['created_at'] as String,
    );
  }
}

class WeightListResult {
  final List<WeightRecord> weights;
  final int total;
  final int page;
  final int pageSize;
  final int totalPages;

  const WeightListResult({
    required this.weights,
    required this.total,
    required this.page,
    required this.pageSize,
    required this.totalPages,
  });

  factory WeightListResult.fromJson(Map<String, dynamic> json) {
    return WeightListResult(
      weights: (json['weights'] as List<dynamic>)
          .map((e) => WeightRecord.fromJson(e as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int,
      page: json['page'] as int,
      pageSize: json['page_size'] as int,
      totalPages: json['total_pages'] as int,
    );
  }
}

enum DewormingTypeE { internal, external, combined }

extension DewormingTypeX on DewormingTypeE {
  String get apiValue {
    switch (this) {
      case DewormingTypeE.internal:
        return 'internal';
      case DewormingTypeE.external:
        return 'external';
      case DewormingTypeE.combined:
        return 'combined';
    }
  }

  String get label {
    switch (this) {
      case DewormingTypeE.internal:
        return '内驱';
      case DewormingTypeE.external:
        return '外驱';
      case DewormingTypeE.combined:
        return '内外同驱';
    }
  }

  static DewormingTypeE fromString(String raw) {
    switch (raw) {
      case 'internal':
        return DewormingTypeE.internal;
      case 'external':
        return DewormingTypeE.external;
      case 'combined':
        return DewormingTypeE.combined;
      default:
        throw ArgumentError('Unknown deworming type: $raw');
    }
  }
}

class DewormingRecord {
  final int id;
  final int petId;
  final int userId;
  final DewormingTypeE dewormingType;
  final String dewormedAt;
  final String createdAt;

  const DewormingRecord({
    required this.id,
    required this.petId,
    required this.userId,
    required this.dewormingType,
    required this.dewormedAt,
    required this.createdAt,
  });

  factory DewormingRecord.fromJson(Map<String, dynamic> json) {
    return DewormingRecord(
      id: json['id'] as int,
      petId: json['pet_id'] as int,
      userId: json['user_id'] as int,
      dewormingType: DewormingTypeX.fromString(json['deworming_type'] as String),
      dewormedAt: json['dewormed_at'] as String,
      createdAt: json['created_at'] as String,
    );
  }
}

class DewormingListResult {
  final List<DewormingRecord> dewormings;
  final int total;
  final int page;
  final int pageSize;
  final int totalPages;

  const DewormingListResult({
    required this.dewormings,
    required this.total,
    required this.page,
    required this.pageSize,
    required this.totalPages,
  });

  factory DewormingListResult.fromJson(Map<String, dynamic> json) {
    return DewormingListResult(
      dewormings: (json['dewormings'] as List<dynamic>)
          .map((e) => DewormingRecord.fromJson(e as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int,
      page: json['page'] as int,
      pageSize: json['page_size'] as int,
      totalPages: json['total_pages'] as int,
    );
  }
}

class DewormingCycleConfig {
  final int? internalCycleDays;
  final int? externalCycleDays;
  final int? combinedCycleDays;
  final bool internalReminderEnabled;
  final bool externalReminderEnabled;
  final bool combinedReminderEnabled;

  const DewormingCycleConfig({
    this.internalCycleDays,
    this.externalCycleDays,
    this.combinedCycleDays,
    required this.internalReminderEnabled,
    required this.externalReminderEnabled,
    required this.combinedReminderEnabled,
  });

  factory DewormingCycleConfig.fromJson(Map<String, dynamic> json) {
    return DewormingCycleConfig(
      internalCycleDays: json['internal_cycle_days'] as int?,
      externalCycleDays: json['external_cycle_days'] as int?,
      combinedCycleDays: json['combined_cycle_days'] as int?,
      internalReminderEnabled: json['internal_reminder_enabled'] as bool,
      externalReminderEnabled: json['external_reminder_enabled'] as bool,
      combinedReminderEnabled: json['combined_reminder_enabled'] as bool,
    );
  }
}

class DewormingStatusItem {
  final bool reminderEnabled;
  final String? lastDewormedAt;
  final int? cycleDays;
  final String? nextDueAt;
  final int? daysRemaining;
  final bool? isOverdue;

  const DewormingStatusItem({
    required this.reminderEnabled,
    this.lastDewormedAt,
    this.cycleDays,
    this.nextDueAt,
    this.daysRemaining,
    this.isOverdue,
  });

  factory DewormingStatusItem.fromJson(Map<String, dynamic> json) {
    return DewormingStatusItem(
      reminderEnabled: json['reminder_enabled'] as bool,
      lastDewormedAt: json['last_dewormed_at'] as String?,
      cycleDays: json['cycle_days'] as int?,
      nextDueAt: json['next_due_at'] as String?,
      daysRemaining: json['days_remaining'] as int?,
      isOverdue: json['is_overdue'] as bool?,
    );
  }
}

class DewormingStatus {
  final DewormingStatusItem internal;
  final DewormingStatusItem external;
  final DewormingStatusItem combined;

  const DewormingStatus({
    required this.internal,
    required this.external,
    required this.combined,
  });

  factory DewormingStatus.fromJson(Map<String, dynamic> json) {
    return DewormingStatus(
      internal: DewormingStatusItem.fromJson(json['internal'] as Map<String, dynamic>),
      external: DewormingStatusItem.fromJson(json['external'] as Map<String, dynamic>),
      combined: DewormingStatusItem.fromJson(json['combined'] as Map<String, dynamic>),
    );
  }

  DewormingStatusItem itemOf(DewormingTypeE type) {
    switch (type) {
      case DewormingTypeE.internal:
        return internal;
      case DewormingTypeE.external:
        return external;
      case DewormingTypeE.combined:
        return combined;
    }
  }
}

class VaccinationRecord {
  final int id;
  final int petId;
  final int userId;
  final String vaccineType;
  final String vaccinatedAt;
  final String createdAt;

  const VaccinationRecord({
    required this.id,
    required this.petId,
    required this.userId,
    required this.vaccineType,
    required this.vaccinatedAt,
    required this.createdAt,
  });

  factory VaccinationRecord.fromJson(Map<String, dynamic> json) {
    return VaccinationRecord(
      id: json['id'] as int,
      petId: json['pet_id'] as int,
      userId: json['user_id'] as int,
      vaccineType: json['vaccine_type'] as String,
      vaccinatedAt: json['vaccinated_at'] as String,
      createdAt: json['created_at'] as String,
    );
  }
}

class VaccinationListResult {
  final List<VaccinationRecord> vaccinations;
  final int total;
  final int page;
  final int pageSize;
  final int totalPages;

  const VaccinationListResult({
    required this.vaccinations,
    required this.total,
    required this.page,
    required this.pageSize,
    required this.totalPages,
  });

  factory VaccinationListResult.fromJson(Map<String, dynamic> json) {
    return VaccinationListResult(
      vaccinations: (json['vaccinations'] as List<dynamic>)
          .map((e) => VaccinationRecord.fromJson(e as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int,
      page: json['page'] as int,
      pageSize: json['page_size'] as int,
      totalPages: json['total_pages'] as int,
    );
  }
}
